File: python/logicprogram.py
#!/usr/bin/python

import itertools

import pandas as pd

import random

# Logic program data structure
# Manage logic rules, matching, generating data, etc

#-------------------------------------------------------------------------------
class Logic_Rule:

    #---------------------------------------------------------------------------
    def __init__(self, head, body):

        self.head = head
        self.body = body

        return

    #---------------------------------------------------------------------------
    def match(self, state) -> bool:

        # returns trud iif the body of the rule mathces the input state

        res = True

        ind_body = 0
        while res and ind_body < len(self.body):

            atom = self.body[ind_body]
            if state[atom[0]] != atom[1]:
                res = False
            else:
                ind_body += 1

        return res

    #---------------------------------------------------------------------------
    def to_string(self) -> str:
        res = str(self.head[0]) + ':ind_' + str(self.head[1]) + ' :-'
        for atom in self.body:
            res += ' ' + str(atom[0]) + ':ind_' + str(atom[1])
        return res

#-------------------------------------------------------------------------------
class Logic_Program:

    #---------------------------------------------------------------------------
    def __init__(self):

        self._variables = [] # labels for each variable
        self.variable_index = {}
        self._values = [] # number of discrete values for each variables

        self._atoms = [] # list of atoms: variables and values
        self._atom_indexes = {} # association list atoms to indexes

        self.rules = []

        return

    #---------------------------------------------------------------------------
    def to_string(self) -> str:
        res = ''
        for rule in self.rules:
            res += self._variables[rule.head[0]] + '_' + str(rule.head[1])
            res += ' :-'
            for elt in rule.body:
                res += ' ' + self._variables[elt[0]] + '_' + str(elt[1])
            res += '.\n'
        return res

    #---------------------------------------------------------------------------
    def add_rule(self, rule):
        self.rules.append(rule)
        return

    #---------------------------------------------------------------------------
    def get_atom_index(self, atom) -> int:
        return self._atom_indexes[atom]

    #---------------------------------------------------------------------------
    def export_to_file(self, file_name):

        file = open(file_name, 'w')

        # write the variables
        for ind_var in range(len(self._variables)):
            line = 'VAR ' + self._variables[ind_var]
            for val in range(self._values[ind_var]):
                line += ' ' + str(val)
            line += '\n'
            file.write(line)

        file.write('\n')

        # write the rules
        for rule in self.rules:

            line = self._variables[rule.head[0]]
            line += '(' + str(rule.head[1]) + ',T) :- '

            for ind_atom in range(len(rule.body)):
                atom = rule.body[ind_atom]
                line += self._variables[atom[0]] + '(' + str(atom[1]) + ',T-1)'
                if ind_atom < len(rule.body)-1:
                    line += ', '

            line += '.\n'

            file.write(line)

        file.close()

        return

    #---------------------------------------------------------------------------
    def generate_all_transitions(self):

        # generate all possible transitions in synchronous semantics

        # generate the set of discrete values for all variables
        values = [ list(range(self._values[var_ind])) for var_ind in range(len(self._variables))]
        # print(values)

        # geenrate the set of all possible states (!exponential)
        states = list(itertools.product(*values))
        # print(states)

        transitions = []

        for state in states:

            next = [[] for _ in range(len(self._variables))]

            # look for all possible conclusions from this state
            for rule in self.rules:

                if rule.match(state):
                    next[rule.head[0]].append(rule.head[1])

            # make sure the state is not a dead end
            dead = False
            for val in next:
                if len(val) == 0:
                    dead = True

            # if dead:
                # print(':( no possible evolution)')

            if not dead: # if there are som successors, make the transitions
                successors = itertools.product(*next)
                for suc in successors:
                    transitions.append((state, suc))

        return transitions

    #---------------------------------------------------------------------------
    def generate_all_transitions_csv_synchrone(self, file_name):

        # generate all possible transitions in synchronous semantics
        # returns the transitions as csv

        # 2 files:  - states of the system observed
        #           - transitions between stats

        # generate the set of discrete values for all variables
        values = [ list(range(self._values[var_ind])) for var_ind in range(len(self._variables))]
        # print(values)

        # geenrate the set of all possible states (!exponential)
        states = list(itertools.product(*values))
        # print(states)

        state_indexes = {}
        for ind in range(len(states)):
            state_indexes[states[ind]] = ind

        transitions = []

        for ind_state in range(len(states)):

            state = states[ind_state]

            next = [[] for _ in range(len(self._variables))]

            # look for all possible conclusions from this state
            for rule in self.rules:

                if rule.match(state) and not rule.head[1] in next[rule.head[0]]:
                    next[rule.head[0]].append(rule.head[1])



            # make sure the state is not a dead end
            dead = False
            for val in next:
                if len(val) == 0:
                    dead = True

            # if dead:
                # print(':( no possible evolution)')

            if not dead: # if there are som successors, make the transitions
                successors = itertools.product(*next)
                for suc in successors:
                    transitions.append((ind_state, state_indexes[suc]))

        # print('transitions: ', transitions)

        states_labels = ['s' + str(ind) for ind in range(len(states))]

        # create the states dataframe
        states_df = pd.DataFrame(states, index = states_labels, columns = self._variables)

        # create the transitions dataframe
        # print(states_df)
        states_df.to_csv(file_name + '_states.csv')

        transitions_labels = ['tr_' + str(ind) for ind in range(len(transitions))]

        transitions_map = [[states_labels[elt[0]],states_labels[elt[1]]] for elt in transitions]

        transitions_df = pd.DataFrame(transitions_map, index = transitions_labels, columns = ['T-1', 'T'])
        transitions_df.to_csv(file_name + '_transitions.csv')

        # print(transitions_df)

        return

    #---------------------------------------------------------------------------
    def generate_all_transitions_csv_asynchrone(self, file_name):

        # generate all possible transitions in synchronous semantics
        # returns the transitions as csv

        # 2 files:  - states of the system observed
        #           - transitions between stats

        # generate the set of discrete values for all variables
        values = [ list(range(self._values[var_ind])) for var_ind in range(len(self._variables))]
        # print(values)

        # geenrate the set of all possible states (!exponential)
        states = list(itertools.product(*values))
        # print(states)

        state_indexes = {}
        for ind in range(len(states)):
            state_indexes[states[ind]] = ind

        # random.shuffle(states)
        # states = states[:int(len(states)/2)]


        transitions = []

        for ind_state in range(len(states)):

            state = states[ind_state]

            next = [[] for _ in range(len(self._variables))]

            # look for all possible conclusions from this state
            for rule in self.rules:

                if rule.match(state) and not rule.head[1] in next[rule.head[0]]:
                    next[rule.head[0]].append(rule.head[1])


            # make sure the state is not a dead end
            dead = False
            ind_var = 0
            for values in next:
                if len(values) == 0:
                    dead = True
                else: # generate successors
                    for value in values:
                        if value != states[ind_state][ind_var]:
                            suc = list(states[ind_state])
                            suc[ind_var] = value
                            suc = tuple(suc)
                            transitions.append((ind_state, state_indexes[suc]))
                ind_var += 1

        # print('transitions: ', transitions)

        # pick a subset of transitions
        random.shuffle(transitions)
        transitions = transitions[:2000]

        states_indexes = [elt[i] for i in [0,1] for elt in transitions]
        states_indexes = list(set(states_indexes))
        print(len(states_indexes))

        states_labels = ['s' + str(ind) for ind in range(len(states))]

        # select subset of states from picked transitions

        # create the states dataframe
        states_df = pd.DataFrame([states[ind] for ind in states_indexes], index = [states_labels[ind] for ind in states_indexes], columns = self._variables)

        # create the transitions dataframe
        # print(states_df)
        states_df.to_csv(file_name + '_states.csv')

        transitions_labels = ['tr_' + str(ind) for ind in range(len(transitions))]

        transitions_map = [[states_labels[elt[0]],states_labels[elt[1]]] for elt in transitions]

        transitions_df = pd.DataFrame(transitions_map, index = transitions_labels, columns = ['T-1', 'T'])
        transitions_df.to_csv(file_name + '_transitions.csv')

        # print(transitions_df)

        return

    #---------------------------------------------------------------------------
    @staticmethod
    def create_from_dataframe(dataframe):

        # create an empty program based on variables from a state dataframe

        lp = Logic_Program()

        lp._variables = dataframe.columns.copy()
        for index in range(len(lp._variables)):

            feature = lp._variables[index]

            lp.variable_index[feature] = index

            values = dataframe[feature].unique()
            # values.sort()

            lp._values.append(len(values))

            for val in values:
                lp._atoms.append((feature, val))
                lp._atom_indexes[(feature, val)] = len(lp._atoms)-1

        return lp

    #---------------------------------------------------------------------------
    @staticmethod
    def load_from_file(file_name):

        lp = Logic_Program()

        file = open(file_name, 'r')
        content = file.read()
        content = content.splitlines()
        file.close()

        # print(content)

        for line in content:

            tokens = line.split(' ')

            # read the variables of the program
            if tokens[0] == 'VAR':
                lp._variables.append(tokens[1])
                lp._values.append(len(tokens[2:]))
                lp.variable_index[lp._variables[-1]] = len(lp._variables)-1

                # red the values
                for value in range(lp._values[-1]):
                    lp._atoms.append((lp._variables[-1], value))
                    lp._atom_indexes[lp._atoms[-1]] = len(lp._atoms)-1

            elif tokens[0] != '': # read the rules

                head = Logic_Program.extract_atom(tokens[0])
                head = (lp.variable_index[head[0]], head[1])

                body = []

                # extract all body atoms
                for token in tokens[2:]:
                    atom = Logic_Program.extract_atom(token)
                    body.append((lp.variable_index[atom[0]], atom[1]))

                lp.rules.append(Logic_Rule(head, body))

        # print(lp.to_string())

        return lp

    #---------------------------------------------------------------------------
    @staticmethod
    def extract_atom(token):


        ind1 = token.index('(')
        ind2 = token.index(',')

        ind3 = token.index('T', ind1)
        ind4 = token.index(')')

        var_label = token[0:ind1]

        var_value = int(token[ind1+1:ind2])

        delay = 0
        if ind3+1 < ind4:
            delay = int(token[ind3+1:ind4])

        return (var_label, var_value, delay)

File: c++/src/LogicProgram.cpp
/*!
 * \file LogicProgram.cpp
 * \author S. Buchet
 * \brief implementation of class LogicProgram
 */

#include "LogicProgram.hpp"
#include "Utils.hpp"

#include <fstream>
#include <iostream>

using namespace std;

/*----------------------------------------------------------------------------*/
LogicProgram::LogicProgram() {

}

/*----------------------------------------------------------------------------*/
uint LogicProgram::nVariables() const {
  return static_cast<uint>(_variables.size());
}

/*----------------------------------------------------------------------------*/
string LogicProgram::varName(uint index) const {
  return _variables[index];
}

/*----------------------------------------------------------------------------*/
uint LogicProgram::varIndex(string varName) const {
  return _varIndex.at(varName);
}

/*----------------------------------------------------------------------------*/
uint LogicProgram::nValues(uint varIndex) const {
  return _nValues[varIndex];
}

/*----------------------------------------------------------------------------*/
void LogicProgram::create(vector<string>& variables, uint nValues) {
  _nValues.resize(variables.size(), nValues);
  _variables = variables;
  _ruleIndexes.resize(variables.size());
  /* variables indexes */
  for(uint ind = 0; ind < _variables.size(); ind ++) {
    _varIndex.insert(pair<string,uint>(_variables[ind], ind));
    _ruleIndexes[ind].resize(nValues);
  }

}

/*----------------------------------------------------------------------------*/
void LogicProgram::addRule(Atom& head, vector<Atom>& body) {
  _ruleIndexes[head.first][head.second].push_back(static_cast<uint>(_rules.size()));
  _rules.push_back(Rule(head, body));
}

/*----------------------------------------------------------------------------*/
void LogicProgram::addVariable(string varName, uint nValues) {
  _variables.push_back(varName);
  _nValues.push_back(nValues);
  _varIndex.insert(pair<string, uint>(varName, _variables.size()-1));
  _ruleIndexes.push_back(vector<vector<uint>>(nValues));
}

/*----------------------------------------------------------------------------*/
void LogicProgram::loadFromFile(string filename) {

  cout << "loading logic program " << filename << endl;

  ifstream file(filename);

  if(file) {

      string line;

      vector<string> tokens;

      while(std::getline(file, line)) {

        tokens.clear();

        if(line.find("VAR") != string::npos) {
          // cout << line << endl;
          Utils::getInstance().split(line, " ",tokens);

          /* variable name */
          _variables.push_back(tokens[1]);
          _nValues.push_back(static_cast<uint>(tokens.size())-2);
          _varIndex.insert(pair<string, uint>(tokens[1], _variables.size()-1));

          /* prepare rule indexes vector */
          _ruleIndexes.push_back(vector<vector<uint>>(_nValues.back()));

          /* variable values */
          // for(uint ind = 2; ind < tokens.size(); ind ++) {
          // }

        } else if(line.find(":-") != string::npos) {

          Utils::getInstance().split(line, " ",tokens);

          Atom head = extractAtom(tokens[0]);

          vector<Atom> body;

          if(tokens[2].size() > 1) {
            for(uint ind = 2; ind < tokens.size(); ind ++) {
              body.push_back(extractAtom(tokens[ind]));
            }
          }

          _ruleIndexes[head.first][head.second].push_back(static_cast<uint>(_rules.size()));
          _rules.push_back(Rule(head, body));

        }


      }

      file.close();
  }

}

/*----------------------------------------------------------------------------*/
void LogicProgram::exportToFile(string fileName) {

  ofstream file(fileName);

  if(file) {

    /* write variables names and values */
    for(uint indVar = 0; indVar < _variables.size(); indVar ++) {
      file << "VAR " << _variables[indVar];
      for(uint val = 0; val < _nValues[indVar]; val ++) {
        file << " " << to_string(val);
      }
      file << endl;
    }

    file << endl;

    /* write rules */
    for(auto& rule: _rules) {

      /* rule head */
      file << _variables[rule.getHead().first] << "(" << rule.getHead().second << ",T)" << " :- ";

      /* rule body */
      for(uint indBody = 0; indBody < rule.bodyLength(); indBody ++) {
        const auto& atom = rule.getBody(indBody);
        file << _variables[atom.first] << "(" << to_string(atom.second) << ",T-1)";
        if(indBody < rule.bodyLength()-1) {
          file << ", ";
        }
      }
      file << "." << endl;

    }

    file.close();

    cout << "writing done!" << endl;

  } else {
    cout << "error" << endl;
  }

}


/*----------------------------------------------------------------------------*/
void LogicProgram::createSubProgram(const LogicProgram& program, uint nVar) {

  vector<bool> selected(program.nVariables(), false);
  for(uint ind = 0; ind < nVar; ind ++) {
    selected[ind] = true;
  }

  _variables.resize(nVar);
  _nValues.resize(nVar);
  for(uint ind = 0; ind < nVar; ind ++) {
    _variables[ind] = program._variables[ind];
    _nValues[ind] = program._nValues[ind];
    _varIndex[_variables[ind]] = ind;
  }

  /* only keep rules for these variables, and make sur there is no condition from another variable */
  for(auto rule : program._rules) {

    if(selected[rule.getHead().first]) { /* make sure this is a selected variable */
      auto body = vector<Atom>();

      /* create a new body containing only selected variables */
      for(auto atom : rule.getBody()) {
        if(selected[atom.first]) {
          body.push_back(atom);
        }
      }

      /* add the new rule */
      _rules.push_back(Rule(rule.getHead(), body));
    }

  }

  /* simplify the new rules */
  simplify();

}



/*----------------------------------------------------------------------------*/
void LogicProgram::getConclusions(const State& state, vector<vector<uint>>& atoms) const {
  atoms.resize(nVariables());
  for(const Rule& rule : _rules) {
    if(rule.match(state)) {
      auto head = rule.getHead();
      vector<uint>& list = atoms.at(head.first);
      if(find(list.begin(), list.end(), head.second) == list.end()) {
        list.push_back(head.second);
      }
    }
  }
}

/*----------------------------------------------------------------------------*/
void LogicProgram::simplify() {

  vector<bool> to_keep(_rules.size(), true);

  for(uint indRule = 0; indRule < _rules.size(); indRule ++) {
    uint indRule2 = 0;
    while(indRule2 < _rules.size() && to_keep[indRule]) {
      if(indRule != indRule2 && _rules[indRule2].subsumes(_rules[indRule])) {
        /* if the rules are equal, decide on indexes which one to keep */
        if(_rules[indRule2].bodyLength() != _rules[indRule].bodyLength() || indRule2 < indRule) {
          to_keep[indRule] = false;
        }
      }
      indRule2 += 1;
    }
  }

  /* keep only the selected rules */
  vector<Rule> newRules;
  for(uint ind = 0; ind < _rules.size(); ind ++) {
    if(to_keep[ind]) {
      newRules.push_back(_rules[ind]);
    }
  }

  _rules.clear();
  _rules.insert(_rules.end(), newRules.begin(), newRules.end());


}

/*----------------------------------------------------------------------------*/
string LogicProgram::toString() const {
  string res;

  for(uint indRule = 0; indRule < _rules.size(); indRule ++) {

    const Rule& rule = _rules[indRule];

    res += _variables[rule.getHead().first] + "_" + to_string(rule.getHead().second);
    res += " :-";

    for(uint ind = 0; ind < rule.bodyLength(); ind ++) {
      res += " " + _variables[rule.getBody(ind).first] + "_" + to_string(rule.getBody(ind).second);
    }

    res += "\n";

  }

  return res;
}


/*----------------------------------------------------------------------------*/
pair<uint,uint> LogicProgram::extractAtom(string str) {
  pair<uint,uint> res;
  uint ind1 = static_cast<uint>(str.find("("));
  res.first = _varIndex[str.substr(0,ind1)];
  uint ind2 = static_cast<uint>(str.find(","));
  res.second = std::atoi(str.substr(ind1+1, ind2-ind1).c_str());
  return res;
}
